src/ImGui.NET.FNA/ImGuiRenderer.cs

```diff
@@ -0,0 +1,313 @@
+            io.DisplayFramebufferScale = new System.Numerics.Vector2(1f, 1f);
+
+            io.MousePosition = new System.Numerics.Vector2(mouse.X, mouse.Y);
+
+            io.MouseDown[0] = mouse.LeftButton == ButtonState.Pressed;
+            io.MouseDown[1] = mouse.RightButton == ButtonState.Pressed;
+            io.MouseDown[2] = mouse.MiddleButton == ButtonState.Pressed;
+
+            var scrollDelta = mouse.ScrollWheelValue - _scrollWheelValue;
+            io.MouseWheel = scrollDelta > 0 ? 1 : scrollDelta < 0 ? -1 : 0;
+            _scrollWheelValue = mouse.ScrollWheelValue;
+        }
+
+        #endregion Setup & Update
+
+        #region Internals
+
+        /// <summary>
+        /// Gets the geometry as set up by ImGui and sends it to the graphics device
+        /// </summary>
+        private unsafe void RenderDrawData(DrawData* drawData)
+        {
+            // Setup render state: alpha-blending enabled, no face culling, no depth testing, scissor enabled, vertex/texcoord/color pointers
+            var lastViewport = _graphicsDevice.Viewport;
+            var lastScissorBox = _graphicsDevice.ScissorRectangle;
+
+            _graphicsDevice.BlendFactor = Color.White;
+            _graphicsDevice.BlendState = BlendState.NonPremultiplied;
+            _graphicsDevice.RasterizerState = _rasterizerState;
+            _graphicsDevice.DepthStencilState = DepthStencilState.DepthRead;
+
+            // Handle cases of screen coordinates != from framebuffer coordinates (e.g. retina displays)
+            ImGui.ScaleClipRects(drawData, ImGui.GetIO().DisplayFramebufferScale);
+
+            // Setup projection
+            _graphicsDevice.Viewport = new Viewport(0, 0, _graphicsDevice.PresentationParameters.BackBufferWidth, _graphicsDevice.PresentationParameters.BackBufferHeight);
+
+            // Render command lists
+            for (int n = 0; n < drawData->CmdListsCount; n++)
+            {
+                var cmdList = drawData->CmdLists[n];
+
+                var vtxBuffer = new ImVector<DrawVertXNA>(&cmdList->VtxBuffer);
+                var vtxArray = new DrawVertXNA[vtxBuffer.Size];
+                for (int i = 0; i < vtxBuffer.Size; i++)
+                {
+                    vtxArray[i] = vtxBuffer[i];
+                }
+
+                var idxBuffer = new ImVector<short>(&cmdList->IdxBuffer);
+
+                uint offset = 0;
+                for (int cmdi = 0; cmdi < cmdList->CmdBuffer.Size; cmdi++)
+                {
+                    var pcmd = &(((DrawCmd*)cmdList->CmdBuffer.Data)[cmdi]);
+                    if (pcmd->UserCallback != IntPtr.Zero) throw new NotSupportedException("The user callback of the command buffer is not supported");
+
+                    // Instead of uploading the complete idxBuffer again and again, just upload what's required
+                    var idxArray = new short[pcmd->ElemCount];
```

Comment:
Again, this will produce lots of garbage every frame. Fixing this one will be a bit trickier because there can be lots of separate draw submissions.